components/brave_viewer/browser/content/brave_viewer_tab_helper.cc

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_viewer/browser/content/brave_viewer_tab_helper.h"
+
+#include <string>
+#include <utility>
+
+#include "base/functional/bind.h"
+#include "base/strings/utf_string_conversions.h"
+#include "base/values.h"
+#include "brave/components/brave_viewer/browser/core/brave_viewer_service.h"
+#include "brave/components/brave_viewer/common/features.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/web_contents.h"
+#include "mojo/public/cpp/bindings/associated_remote.h"
+#include "net/base/registry_controlled_domains/registry_controlled_domain.h"
+#include "third_party/blink/public/common/associated_interfaces/associated_interface_provider.h"
+
+namespace brave_viewer {
+
+// static
+void BraveViewerTabHelper::MaybeCreateForWebContents(
+    content::WebContents* contents,
+    const int32_t world_id) {
+  if (!base::FeatureList::IsEnabled(brave_viewer::features::kBraveViewer)) {
+    return;
+  }
+
+  if (!SameDomainOrHost(
+          contents->GetLastCommittedURL(),
+          url::Origin::CreateFromNormalizedTuple("https", "youtube.com", 443),
+          net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES)) {
+    return;
+  }
+
+  brave_viewer::BraveViewerTabHelper::CreateForWebContents(contents, world_id);
+}
+
+BraveViewerTabHelper::BraveViewerTabHelper(content::WebContents* web_contents,
+                                           const int32_t world_id)
+    : WebContentsObserver(web_contents),
+      content::WebContentsUserData<BraveViewerTabHelper>(*web_contents),
+      world_id_(world_id),
+      brave_viewer_service_(BraveViewerService::GetInstance()) {
+  DCHECK(brave_viewer_service_);
+}
+
+BraveViewerTabHelper::~BraveViewerTabHelper() = default;
+
+void BraveViewerTabHelper::OnTestScriptResult(
+    const content::GlobalRenderFrameHostId& render_frame_host_id,
+    base::Value value) {
+  if (value.GetIfBool().value_or(false)) {
+    // TODO check state and/or trigger Brave Viewer dialog here
+  }
```

Comment:
Still need to write tests for this PR, but this should be the appropriate place to hook in for the platform-specific native dialog UI.